grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceMigrationService.groovy

```diff
@@ -0,0 +1,108 @@
+     * feature was introduced.
+     */
+    void updateAmountFieldForPrepaymentInvoiceItems() {
+        InvoiceType prepaymentInvoiceType = InvoiceType.findByCode(InvoiceTypeCode.PREPAYMENT_INVOICE)
+
+        Invoice.findAllByInvoiceType(prepaymentInvoiceType).each { Invoice prepaymentInvoice ->
+            // The amount field will be null for all pre-existing invoices since the field was not being used.
+            prepaymentInvoice.invoiceItems.find { it.amount == null }.each { InvoiceItem prepaymentInvoiceItem ->
+                prepaymentInvoiceItem.amount = computePrepaymentInvoiceItemAmount(prepaymentInvoiceItem)
+                prepaymentInvoiceItem.save(failOnError: true, flush: true)
+            }
+        }
+    }
+
+    // TODO: Can we merge this with PrepaymentInvoiceService? That way if the amount calculation ever changes, it will get picked up automatically here.
+    private BigDecimal computePrepaymentInvoiceItemAmount(InvoiceItem prepaymentInvoiceItem) {
+        BigDecimal prepaymentPercent = (prepaymentInvoiceItem.order.paymentTerm?.prepaymentPercent ?: Constants.DEFAULT_PAYMENT_PERCENT) / 100
+
+        // If the invoice item is for an adjustment.
+        OrderAdjustment orderAdjustment = prepaymentInvoiceItem.orderAdjustment
+        if (orderAdjustment) {
+            return (orderAdjustment?.canceled ? 0 : orderAdjustment.totalAdjustments) * prepaymentPercent
+        }
+
+        // Else if the invoice item is for a canceled order item.
+        OrderItem orderItem = prepaymentInvoiceItem.orderItem
+        if (orderItem.orderItemStatusCode == OrderItemStatusCode.CANCELED) {
+            return 0
+        }
+
+        // Else the invoice item is for a non-canceled order item.
+        return (orderItem.quantity ?: 0) * (orderItem.unitPrice ?: 0.0) * prepaymentPercent
+    }
+
+    /**
+     * For STEP 2 of the migration.
+     *
+     * For every item in a given prepayment invoice, computes the inverse invoice item and insert it to the database
+     * by adding it to the given final invoice.
+     *
+     * This method assumes that because there is both a prepayment invoice and a final invoice, that the order
+     * associated with both invoices has been fully invoiced, and so there's no need to check each item of the
+     * prepayment invoice individually, they should ALL need to have inverse items created for them.
+     */
+    void generateInverseInvoiceItems(Invoice prepaymentInvoice, Invoice finalInvoice) {
+        prepaymentInvoice.invoiceItems.each { InvoiceItem prepaymentInvoiceItem ->
+            List<InvoiceItem> inverseItems = createInverseInvoiceItems(prepaymentInvoiceItem, finalInvoice)
+            inverseItems?.each { finalInvoice.addToInvoiceItems(it) }
+        }
+        finalInvoice.save(failOnError: true, flush: true)
+    }
+
+    private List<InvoiceItem> createInverseInvoiceItems(InvoiceItem prepaymentInvoiceItem, Invoice finalInvoice) {
```

Comment:
I **think** it should be just a clone of prepaymentInvoiceItem with modified quantity and amount according to needs